lib/ocpp/v16/profile.cpp

```diff
@@ -0,0 +1,600 @@
+    ocpp::v16::DEFAULT_AND_MAX_NUMBER_PHASES; // also the maximum number of phases
+constexpr float low_voltage = 230;            // in Volts
+
+using ocpp::v16::EnhancedChargingSchedulePeriod;
+
+inline std::int32_t elapsed_seconds(const ocpp::DateTime& to, const ocpp::DateTime& from) {
+
+    return duration_cast<seconds>(to.to_time_point() - from.to_time_point()).count();
+}
+
+inline ocpp::DateTime floor_seconds(const ocpp::DateTime& dt) {
+    return ocpp::DateTime(std::chrono::floor<seconds>(dt.to_time_point()));
+}
+
+void update_itt(int duration, std::vector<EnhancedChargingSchedulePeriod>::const_iterator& itt,
+                const std::vector<EnhancedChargingSchedulePeriod>::const_iterator& end,
+                EnhancedChargingSchedulePeriod& period, int& period_duration) {
+    if (itt != end) {
+        auto next = std::next(itt);
+        period = *itt;
+        period_duration = (next != end) ? next->startPeriod : std::numeric_limits<int>::max();
+
+        if (duration >= period_duration) {
+            itt++;
+            if (itt != end) {
+                period = *itt;
+                period_duration = (next != end) ? next->startPeriod : std::numeric_limits<int>::max();
+            }
+        }
+    }
+
+    if (itt == end) {
+        period.startPeriod = -1;
+        period_duration = std::numeric_limits<int>::max();
+    }
+}
+
+std::pair<float, std::int32_t> convert_limit(const ocpp::v16::period_entry_t* const entry,
+                                             const ocpp::v16::ChargingRateUnit selected_unit) {
+    assert(entry != nullptr);
+    float limit = entry->limit;
+    std::int32_t number_phases = entry->number_phases.value_or(default_number_phases);
+
+    // if the units are the same - don't change the values
+    if (selected_unit != entry->charging_rate_unit) {
+        if (selected_unit == ocpp::v16::ChargingRateUnit::A) {
+            limit = entry->limit / (low_voltage * number_phases);
+        } else {
+            limit = entry->limit * (low_voltage * number_phases);
+        }
+    }
+
+    return {limit, number_phases};
+}
+
+} // namespace
+
+namespace ocpp::v16 {
+
+constexpr bool operator==(const ChargingSchedulePeriod& lhs, const ChargingSchedulePeriod& rhs) {
+    return (lhs.startPeriod == rhs.startPeriod) && (lhs.limit == rhs.limit);
+}
+
+constexpr bool operator!=(const ChargingSchedulePeriod& lhs, const ChargingSchedulePeriod& rhs) {
+    return !(lhs == rhs);
+}
+
+inline std::ostream& operator<<(std::ostream& os, const period_entry_t& entry) {
+    os << entry.start << " " << entry.end << " S:" << entry.stack_level << " " << entry.limit
+       << ((entry.charging_rate_unit == ChargingRateUnit::A) ? "A" : "W");
+    return os;
+}
+
+/**
+ * \brief populate a schedule period
+ * \param in_start the start time of the profile
+ * \param in_duration the time in seconds from the start of the profile to the end of this period
+ * \param in_period the details of this period
+ */
```

Comment:
In libocpp these comment blocks are used for the doxygen. Please apply everywhere
```suggestion
 /// \brief populate a schedule period
 /// \param in_start the start time of the profile
 /// \param in_duration the time in seconds from the start of the profile to the end of this period
 /// \param in_period the details of this period
```